### ISSUE TITLE
[BUG] Cannot authenticate on NixOS via login or npmrc

### ISSUE THREAD

**Issue by jaythomas** [npm/cli#7802]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

I am using [NixOS](https://nixos.org/) with [fnm](https://github.com/Schniz/fnm) and set my authentication tokens via [direnv](https://github.com/direnv/direnv). Typically I can add my private registry via the adduser or login command. This command works with (Arch Linux + fnm + direnv) but not from (NixOS + nix-ld + fnm + direnv):

```
npm adduser --registry https://nexus.XXXXXX/repository/npm-hosted/
npm notice Log in on https://nexus.XXXXXX/repository/npm-hosted/
Username: XXXXXXX
Password:
Email: (this IS public)
⠦

npm error code E401
npm error Unable to authenticate, need: BASIC realm="Sonatype Nexus Repository Manager"
npm error A complete log of this run can be found in: /home/myuser/.npm/_logs/2024-10-01T14_29_32_111Z-debug-0.log
```

I tried manually setting up my registry and token in .npmrc:

```
@myorg:registry=https://nexus.XXXXXX/repository/npm-hosted/
//nexus.XXXXXX/repository/npm-hosted/:_authToken=NpmToken.xxxxxxx-xxxxx-xxxx-xxxx-xxxxxxxxxxxx
```

Again, this method works as expected with (Arch Linux + fnm + direnv) but not via (NixOS + nix-ld + fnm + direnv). I receive the same error via whoami:

```
$ npm whoami --registry https://nexus.XXXXXX/repository/npm-hosted/
npm error code E401
npm error Unable to authenticate, need: BASIC realm="Sonatype Nexus Repository Manager"
npm error A complete log of this run can be found in: /home/myuser/.npm/_logs/2024-10-01T15_46_05_297Z-debug-0.log
```

I can confirm the environment variable is populating so `direnv` is doing its job:

package.json:
```json
{
  ...
  "name": "myrepo",
  "version": "1.12.21",
  "private": true,
  "workspaces": [
    "packages/*"
  ],
  "scripts": {
    "print": "node -e \"console.log(process.env.NPM_TOKEN)\""
  },
  "dependencies": {
  ...
}
```

```
$ NPM_TOKEN=foo npm run print

> myrepo@1.12.21 print
> node -e "console.log(process.env.NPM_TOKEN)"

foo
```

### Expected Behavior

With an appropriately set .npmrc, `npm whoami` should return successfully.

### Steps To Reproduce

1. In this environment...

<details>
<summary>configuration.nix</summary>

```nix
### snip
  users.users.myuser.packages = with pkgs; [ fnm ];

  programs.mtr.enable = true;

  programs.nix-ld = {
    enable = true;
    libraries = with pkgs; [
      curl
      glib
      icu
      libdrm
      libusb1
      libuuid
      nspr
      nss
      openssl
      stdenv.cc.cc
      zlib
    ];
  };

  home-manager.users.myuser.programs.direnv.enable = true;
### snip
```
</details>

2. With this config...

<details>
<summary>.npmrc</summary>

```
@myorg:registry=https://nexus.XXXXXX/repository/npm-hosted/
//nexus.XXXXXX/repository/npm-hosted/:_authToken=NpmToken.xxxxxxx-xxxxx-xxxx-xxxx-xxxxxxxxxxxx
```
</details>

4. Run...

```
npm whoami
npm whoami --registry https://nexus.XXXXXX/repository/npm-hosted/
```

6. See error...

```
$ npm whoami
npm error code ENEEDAUTH
npm error need auth This command requires you to be logged in.
npm error need auth You need to authorize this machine using `npm adduser`
npm error A complete log of this run can be found in: /home/myuser/.npm/_logs/2024-10-01T16_00_18_905Z-debug-0.log
$ npm whoami --registry https://nexus.XXXXXX/repository/npm-hosted/
npm error code E401
npm error Unable to authenticate, need: BASIC realm="Sonatype Nexus Repository Manager"
npm error A complete log of this run can be found in: /home/myuser/.npm/_logs/2024-10-01T16_00_27_244Z-debug-0.log
```

This can be reproduced using bash, fish, or zsh.

### Environment

- npm: 10.8.1
- Node.js: 20.16.0
- OS Name:
```
uname -a
Linux selkie 6.10.5 #1-NixOS SMP PREEMPT_DYNAMIC Wed Aug 14 13:34:38 UTC 2024 x86_64 GNU/Linux
```
- System Model Name: 
```
$ lscpu | grep "Model name"
Model name:                           Intel(R) Core(TM) Ultra 9 185H
```

- npm config:

```ini
; "user" config from /home/myuser/.npmrc

; @explorelearning:registry = "https://nexus.XXXXXX/repository/npm-hosted/" ; overridden by project
; //nexus.XXXXXX/repository/npm-hosted/:_authToken = (protected) ; overridden by project

; "project" config from /home/myuser/code/myrepo/.npmrc

@explorelearning:registry = "https://nexus.XXXXXX/repository/npm-hosted/"
//nexus.XXXXXX/repository/npm-hosted/:_authToken = (protected)

; node bin location = /home/myuser/.local/share/fnm/node-versions/v20.16.0/installation/bin/node
; node version = v20.16.0
; npm local prefix = /home/myuser/code/myrepo
; npm version = 10.8.1
; cwd = /home/myuser/code/myrepo
; HOME = /home/myuser
; Run `npm config ls -l` to show all defaults
```

**Comment by milaninfy** [comment]
This looks like it's specific to NixOS or Nix-ld, if it works for other distros and fails for this one then it's does not looks like issue with npm.

**Comment by jaythomas** [comment]
That's very likely, however I'm not sure what further steps to take to rule out the issue being from npm's side. For instance, if the environment variables npm needs are getting provided to it then what is preventing it from authenticating?

Any advice on what to test from my end to determine whether the issue is from the npm or the nixos side would be appreciated.

**Comment by milaninfy** [comment]
one more thing I can think of is, I am sure if it works on other os then your npmrc is correct  but still you can try 
`npm adduser --registry https://nexus.XXXXXX/repository/npm-hosted/ --scope=<your_scope> `

**Comment by jaythomas** [comment]
Similar results running the command different ways, with or without .npmrc files preset. Unfortunately the logs don't print out the HTTP request details as I'm sure the headers (or lack of) would be telling.

<details>
<summary>npm_debug.log</summary>

```
0 verbose cli /home/myuser/.local/share/fnm/node-versions/v20.16.0/installation/bin/node /run/user/1000/fnm_multishells/5623_1728037881105/bin/npm
1 info using npm@10.8.1
2 info using node@v20.16.0
3 silly config load:file:/home/myuser/.local/share/fnm/node-versions/v20.16.0/installation/lib/node_modules/npm/npmrc
4 silly config load:file:/home/myuser/code/myrepo/.npmrc
5 silly config load:file:/home/myuser/.npmrc
6 silly config load:file:/home/myuser/.local/share/fnm/node-versions/v20.16.0/installation/etc/npmrc
7 verbose title npm adduser
8 verbose argv "adduser" "--registry" "https://nexus.XXXXXX.net/repository/npm-hosted/" "--scope" "@myorg"
9 verbose logfile logs-max:10 dir:/home/myuser/.npm/_logs/2024-10-05T00_28_43_904Z-
10 verbose logfile /home/myuser/.npm/_logs/2024-10-05T00_28_43_904Z-debug-0.log
11 notice Log in on https://nexus.XXXXXX.net/repository/npm-hosted/
12 verbose web adduser before first POST
13 silly logfile start cleaning logs, removing 1 files
14 silly logfile done cleaning log files
15 http fetch POST 401 https://nexus.XXXXXX.net/repository/npm-hosted/-/v1/login 315ms
16 verbose web add user not supported, trying couch
17 verbose adduser before first PUT {
17 verbose adduser   _id: 'org.couchdb.user:myuser',
17 verbose adduser   name: 'myuser',
17 verbose adduser   password: 'XXXXX',
17 verbose adduser   email: 'XXXXXXX.com',
17 verbose adduser   type: 'user',
17 verbose adduser   roles: [],
17 verbose adduser   date: '2024-10-05T00:29:47.277Z'
17 verbose adduser }
18 http fetch PUT 401 https://nexus.XXXXXX.net/repository/npm-hosted/-/user/org.couchdb.user:myuser 185ms
19 verbose stack HttpErrorAuthUnknown: Unable to authenticate, need: BASIC realm="Sonatype Nexus Repository Manager"
19 verbose stack     at /home/myuser/.local/share/fnm/node-versions/v20.16.0/installation/lib/node_modules/npm/node_modules/npm-registry-fetch/lib/check-response.js:80:17
19 verbose stack     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
19 verbose stack     at async putCouch (/home/myuser/.local/share/fnm/node-versions/v20.16.0/installation/lib/node_modules/npm/node_modules/npm-profile/lib/index.js:133:18)
19 verbose stack     at async otplease (/home/myuser/.local/share/fnm/node-versions/v20.16.0/installation/lib/node_modules/npm/lib/utils/auth.js:8:12)
19 verbose stack     at async Object.adduser (/home/myuser/.local/share/fnm/node-versions/v20.16.0/installation/lib/node_modules/npm/lib/utils/auth.js:58:11)
19 verbose stack     at async AddUser.exec (/home/myuser/.local/share/fnm/node-versions/v20.16.0/installation/lib/node_modules/npm/lib/commands/adduser.js:31:35)
19 verbose stack     at async Npm.exec (/home/myuser/.local/share/fnm/node-versions/v20.16.0/installation/lib/node_modules/npm/lib/npm.js:207:9)
19 verbose stack     at async module.exports (/home/myuser/.local/share/fnm/node-versions/v20.16.0/installation/lib/node_modules/npm/lib/cli/entry.js:74:5)
20 verbose statusCode 401
21 verbose pkgid org.couchdb.user:myuser
22 error code E401
23 error Unable to authenticate, need: BASIC realm="Sonatype Nexus Repository Manager"
24 verbose cwd /home/myuser/code/myrepo
25 verbose os Linux 6.10.5
26 verbose node v20.16.0
27 verbose npm  v10.8.1
28 verbose exit 1
29 verbose code 1
30 error A complete log of this run can be found in: /home/myuser/.npm/_logs/2024-10-05T00_28_43_904Z-debug-0.log
```

</details>

**Comment by milaninfy** [comment]
Based on some error logs that I could see in the logs I think it might be worth checking out documentation of the registry that you are using and see if it can be ruled out. 

`npm adduser --auth-type=legacy  <your registry and scope config>`
https://help.sonatype.com/en/npm-security.html

**Comment by jaythomas** [comment]
I took the API key given to me by my private registry when I set up my npm account on Arch Linux and was able to get a working set up on NixOS without logging in. I wish npm debug logs provided more information in the HTTP request so issues like this could be better diagnosed, but the scope of my original problem I am able to work around so I'm going to close this issue. Thank you @milaninfy for assistance.
